### PR TITLE
Add profiles for build-systems (/package-managers)

### PR DESCRIPTION
**Pull request by rusty-snake**
Profiles: bunler, cargo (refactor), cmake (untested), make, meson, pip
All redirect to build-systems-common.profile

Other fixes:
 - blacklist ${HOME}/.bundle
 - blacklist ${HOME}/.cargo/* -> blacklist ${HOME}/.cargo
 - blacklist /usr/lib64/ruby

You can now `firejail make` firejail :laughing:.

### CODE REVIEW

**Review by reinerh — APPROVED** [review]

**Review by kmk3 — COMMENTED** [review]
While language-specific package managers tend to live in their own world,
general-purpose build-systems usually try to integrate with the system, and so
I'd expect more breakage from such profiles.

I'd make a separate PR for make, as it is very general purpose and another for
cmake/meson, as they are similar to autoconf, in the sense that they can be
used to run arbitrary commands to test for and configure things.  If many
people end up using firejail with these tools, I'd expect more corner cases
than usual to be found (i.e.: compared to programs not related to development),
so being able to reference separate PRs on issues could help gauge the results
with timeline events.

File: etc/profile-a-l/build-systems-common.profile [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+include disable-common.inc
+include disable-exec.inc
+include disable-interpreters.inc
+include disable-programs.inc
+include disable-shell.inc
+include disable-X11.inc
+include disable-xdg.inc
+
+whitelist ${HOME}/Projects
```
Comment (kmk3) [inline etc/profile-a-l/build-systems-common.profile:31]: Considering that there is no standard path for this, if the includers of
build-systems-common.profile are ever enabled by default, I'd wager that there
would be many issues opened due to the whitelist above.  I suppose that we
could at least add comments like:

```
# Note: Be sure to whitelist your own development project directories:
#whitelist ${HOME}/YOUR_PROJECT_DIR
```

This is kind of similar to what was added to npm.profile on #3866:

>     # If you want whitelisting, change the line below to your npm projects directory
>     # and uncomment the lines below.
>     [...]
>     #whitelist ${HOME}/Projects

Which was later changed to the following on #3876:

>     # If you want whitelisting, change ${HOME}/Projects below to your npm projects directory
>     # and uncomment the lines below.
>     [...]
>     #whitelist ${HOME}/Projects

(And which now lives on nodejs-common.profile since #4255)

Cc: @aidalgol @glitsj16

File: etc/profile-a-l/bundle.profile [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# Firejail profile for bundle
```
[the remark this answers is not in the document]
Comment (rusty-snake) [reply etc/profile-a-l/bundle.profile:1]: > Wouldn't bundler also need access to the global gem cache or does it always
completely ignore it?

IDK but we can add it.

> So does this profile work with e.g.: make/gcc?
>
> Also, does it work when RVM is used?

Untested (I don't really use ruby except for jekyll)

File: etc/profile-m-z/make.profile [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# Firejail profile for make
```
[the remark this answers is not in the document]
Comment (rusty-snake) [reply etc/profile-m-z/make.profile:1]: > I'd say that make is much more general-purpose compared to the rest, so I'd be
very cautious in adding such a profile.

None of these profile are in firecfg.config (for reasons) they are provided as an starting point for advanced users who want to sandbox make/meson/... and must be customized (via command-line or .locals) in the most cases. Because build-systems can require anything, you would need to allow anything if you want a general profile.

>  It can be used with any tool and any
programming language, so I don't think include disable-interpreters.inc makes
sense for it.

IMHO adding disable-interpreters by default makes sense, if you need to build something that requires let's say python, then `include allow-python3.inc` in `make.local` or run `firejail --include=/etc/firejail/allow-python3.inc --profile=make make <target>`.

File: etc/profile-m-z/meson.profile [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+quiet
+# Persistent local customizations
+include meson.local
+# Persistent global definitions
+include globals.local
+
+# Allow python3 (blacklisted by disable-interpreters.inc)
+include allow-python3.inc
+
+private-bin meson,python3*
```
[the remark this answers is not in the document]
Comment (rusty-snake) [reply etc/profile-m-z/meson.profile:13]: Yes, it does. If you need other programs, `cutom_target`s, ... just add `ignore private-bin`.

smartass OT comment: meson calls them function AFAIK.

File: etc/profile-a-l/cmake.profile [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# Firejail profile for cargo
+# Description: The Rust package manager
+# This file is overwritten after every install/update
+quiet
+# Persistent local customizations
+include cargo.local
+# Persistent global definitions
+include globals.local
+
+private-bin cmake
```
[the remark this answers is not in the document]
Comment (rusty-snake) [reply etc/profile-a-l/cmake.profile:10]: Same as for meson.

File: etc/profile-a-l/build-systems-common.profile (hunk 1 above)
Comment (rusty-snake) [reply etc/profile-a-l/build-systems-common.profile:31]: IMHO we must not ever sandbox build-systems by default as they can require anything.

I don't even sandbox IDEs. The profile are very weak / must be weak (full D-Bus, no whitelisting, minimal blacklisting, ...) the win is very low to nothing but the sandbox regularly breaks things will developing and since nesting isn't supported you can not sandbox the things which really need sandboxing in a tight profile.

Comment (aidalgol) [reply etc/profile-a-l/build-systems-common.profile:31]: I am in favour of changing this to make the whitelisting opt-in just as in `nodejs-common.profile`, and I think `nodejs-common.profile` should be refactored as part of this set of changes in the same way the other package manager profiles have been.

Comment (rusty-snake) [reply etc/profile-a-l/build-systems-common.profile:31]: My idea was that since you already need to type `firejail meson compile -C _builddir` (for example), using `firejail --whitelist=~/my_project_dir meson compile -C _builddir` isn't much more complicated. Anyway, will make whitelisting opt-in.

File: etc/profile-m-z/meson.profile (hunk 4 above)
[the remark this answers is not in the document]
Comment (kmk3) [reply etc/profile-m-z/meson.profile]: > smartass OT comment: meson calls them function AFAIK.

And cmake calls them "commands", but since they're all autoconf-like tools and
since these things are the same as autoconf macros in purpose and function,
it's easier to just call everything macros :p

File: etc/profile-a-l/build-systems-common.profile (hunk 1 above)
Comment (kmk3) [reply etc/profile-a-l/build-systems-common.profile]: @rusty-snake [on Sep 10](https://github.com/netblue30/firejail/pull/4519#discussion_r706252880):

> Anyway, will make whitelisting opt-in.

Thanks.  Can you still put a comment above it noting that it's intended to be
changed?

It's not clear to me that it's supposed to be user-customized; I think that the
path could be easily mistaken as something related to a build system.

**Comment by kmk3** [comment]
@rusty-snake [on Sep 9](https://github.com/netblue30/firejail/pull/4519#discussion_r705609486):

> > I'd say that make is much more general-purpose compared to the rest, so I'd
> > be very cautious in adding such a profile.
> 
> None of these profile are in firecfg.config (for reasons) they are provided
> as an starting point for advanced users who want to sandbox make/meson/...
> and must be customized (via command-line or .locals) in the most cases.
> Because build-systems can require anything, you would need to allow anything
> if you want a general profile.

I think it would be nice to eventually have bundler on firecfg.config (I had
wanted to create a profile for it too).  I also think that it may be possible
to make a generic enough cmake profile for building system packages.

Since I considered these doable (more than for make at least) and since it
wasn't specified that the profiles in this PR are mostly templates, I assumed
that all of them were really intended to be enabled by default in the future.
If they're all intended to be templates, then I'd rather continue this in a
discussion later.

> > It can be used with any tool and any programming language, so I don't think
> > include disable-interpreters.inc makes sense for it.
> 
> IMHO adding disable-interpreters by default makes sense, if you need to build
> something that requires let's say python, then `include allow-python3.inc` in
> `make.local` or run `firejail --include=/etc/firejail/allow-python3.inc
> --profile=make make <target>`.

Same as above.

File: etc/profile-m-z/make.profile (hunk 3 above)
[the remark this answers is not in the document]
Comment (kmk3) [reply etc/profile-m-z/make.profile:1]: Answered here:
<https://github.com/netblue30/firejail/pull/4519#issuecomment-922891679>

Marking as resolved.

**Review by matu3ba — CHANGES_REQUESTED** [review]
Many whitelist stuff is not needed and there are no descriptions how this general profile should be used for the important use cases.
